src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+
+Copyright (c) 2019 Paul-Louis Ageneau
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/config.hpp" // for TORRENT_USE_RTC
+
+#if TORRENT_USE_RTC
+
+#include "libtorrent/websocket_tracker_connection.hpp"
+#include "libtorrent/aux_/escape_string.hpp"
+#include "libtorrent/aux_/session_settings.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/ip_filter.hpp"
+#include "libtorrent/socket.hpp"
+#include "libtorrent/socket_io.hpp"
+#include "libtorrent/tracker_manager.hpp"
+
+#include "nlohmann/json.hpp"
+
+#include <algorithm>
+#include <cctype>
+#include <cstdio> // for snprintf
+#include <functional>
+#include <list>
+#include <locale>
+#include <string>
+#include <vector>
+
+namespace libtorrent {
+
+using namespace std::placeholders;
+
+using json = nlohmann::json;
+
+std::string from_latin1(std::string const& s) {
```

Comment:
This should take a `string_view` instead. But perhaps more importantly, do JSON strings have to be unicode and UTF-8 encoded? And the only reason to have this function is to treat arbitrary bytes as latin-1 and convert it to UTF-8 and back again?

If this conversion is in fact necessary, these functions should probably live in a separate translation unit, have unit tests ans fuzzers.

Reply:
actually, as far as I can tell, this is *only* used to encode binary data into json strings, so the parameter should probably be a `span<char const>`. That way, you wouldn't have to perform the awkward conversions at every call site either

Reply:
Yeah, that's kind of messy. JSON strings have to be UTF-8 encoded, it's required by spec and checked by the json lib. However, the original WebTorrent implementer made the weird choice of using latin1 to UTF-8 encoding to represent binary data. I don't see the point of it, since base64 would have been better, but that's how existing implementations work.

I'll put these functions outside and change the parameter to `span<char const>`.